### DISCUSSION TITLE
Integrate FTP server container

### DISCUSSION THREAD

**Discussion by luca-bassoricci** [quarkusio/quarkus#29270] [Q&A] [answered]
I'm going to integrate an FTP server based on `delfer/alpine-ftp-server` image only for `dev` profile.
Are there any guideline/advice on how to integrate other images than managed ones? 
eg, if devservices are running should I reuse some Docker components exposed by Quarkus?
That's what I wrote until now

```
public class FTPServerContainer extends FixedHostPortGenericContainer<FTPServerContainer>
{
	public FTPServerContainer()
	{
		super(DockerImageName.parse("delfer/alpine-ftp-server:latest").asCanonicalNameString());
		self().withExposedPorts(21).withEnv("MIN_PORT", String.valueOf(21000)).withEnv("MAX_PORT",
				String.valueOf(21000));
		self().withFixedExposedPort(21, 21);
		self().withFixedExposedPort(21000, 21000);
	}
}

@UnlessBuildProfile("prod")
@IfBuildProperty(name = "devservice.ftp.enabled", stringValue = "true", enableIfMissing = true)
@Startup
@ApplicationScoped
@RequiredArgsConstructor
@Slf4j
public class FTPServerDevServiceBean
{
	private final Config config;
	private FTPServerContainer ftp;

	@PostConstruct
	void init()
	{
		final var auth = config.authorization();
		var container = new FTPServerContainer();
		final var envMap = container.getEnvMap();
		container.withCreateContainerCmdModifier(cmd -> cmd.withName("dev-ftp"));
		container.withStartupTimeout(Duration.ofSeconds(10));
		envMap.put("USERS",
				String.join("|", auth.username(), auth.password().orElse(""), this.config.upload().remoteUploadPath()));
		container.start();
		this.ftp = container;
		log.info("DevService FTP server started");
	}

	@PreDestroy
	void destroy()
	{
		if (this.ftp != null)
		{
			this.ftp.stop();
			log.info("DevService FTP server stopped");
		}
	}
}
```
Luca.

**Comment by maxandersen** [comment]
are you doing that in a extension or are you trying to do it in a Quarkus application?

**Reply by luca-bassoricci** [reply]
In a Quarkus app. 
Attached code works fine for our purposes, but I'm not sure if this is the right way to solve the problem.

**Comment by luca-bassoricci** [comment]
up

**Comment by luca-bassoricci** [comment]
up

**Comment by luca-bassoricci** [comment]
up

**Comment by gsmet** [comment, marked as answer]
The problem with your approach is that you will have the Testcontainers dependency in your runtime classpath. It might not be a problem for you though.

Creating an extension would allow you to have the dependency only in the deployment classpath (but you wouldn't be able to start the container outside of dev or test mode).

If you decide to go for the extension (a simple extension is relatively simple to create), you have examples in the codebase on how to start the dev service in dev/test modes.

So, there's no one size fits all, it really depends on what you want to achieve.

**Reply by luca-bassoricci** [reply]
I've approched the problem in this way because I need this ftp server only in dev/test, not during production.
If Docker engine is automatically managed (start/stop/reload etc) by Quarkus I suppose I can stay with my solution.

@gsmet thanks for help
